pkg/ccl/sqlproxyccl/forwarder.go

```diff
@@ -130,37 +260,548 @@ func (f *forwarder) Close() {
+			//   2. The last message sent to the SQL pod was a CopyDone(c), and
+			//      a ReadyForQuery(Z) has already been received at the time of
+			//      evaluation.
+			//   3. The last message sent to the SQL pod was a CopyFail(f), and
+			//      a ReadyForQuery(Z) has already been received at the time of
+			//      evaluation.
+			var isSafeTransferPoint bool
+			switch f.clientMessageTypeSent {
+			case clientMsgAny,
+				pgwirebase.ClientMsgSync,
+				pgwirebase.ClientMsgSimpleQuery,
+				pgwirebase.ClientMsgCopyDone,
+				pgwirebase.ClientMsgCopyFail:
+				if atomic.LoadInt32(&f.atomics.isServerMsgReadyReceived) == 1 {
+					isSafeTransferPoint = true
+				}
+				fallthrough
+			default:
+				if !isSafeTransferPoint {
+					// We have to abort the transfer.
+					f.finishTransfer()
+					continue requestProcessor
+				}
+			}
+
+			// Update the state first so that the response processor could start
+			// processing. If we update the state after sending the request, we
+			// may miss response messages.
+			atomic.StoreInt32(&f.atomics.state, stateTransferInProgress)
+
+			if err := f.sendTransferStateRequest(); err != nil {
+				return err
+			}
+
+			// Wait until transfer is completed.
+			select {
+			case <-f.ctx.Done():
+				return f.ctx.Err()
+			case <-f.transferCh:
+				// Channel is closed whenever transfer completes, so we are done.
+			}
+
+		case stateTransferInProgress:
+			// This cannot happen unless there is a bug. While the transfer is
+			// in progress, the request processor has to be blocked, and the
+			// only way to transition into this state is to go through the
+			// stateTransferRequested state.
+			//
+			// Return an error to close the connection, rather than letting it
+			// continue silently.
+			return errTransferProtocol
+		}
 	}
 	return f.ctx.Err()
 }
 
 // handleServerToClient handles the communication from the server to the client.
 // This returns a context cancellation error whenever the forwarder's context
-// is cancelled, or whenever forwarding fails. When ForwardMsg gets blocked on
-// Read, we will unblock that by closing serverConn through f.Close().
+// is cancelled, or whenever forwarding fails.
 func (f *forwarder) handleServerToClient() error {
 	for f.ctx.Err() == nil {
-		if _, err := f.serverInterceptor.ForwardMsg(f.clientConn); err != nil {
+		typ, _, err := f.serverInterceptor.PeekMsg()
+		if err != nil {
 			return err
 		}
+
+		switch atomic.LoadInt32(&f.atomics.state) {
+		case stateReady, stateTransferRequested:
+			// Have we seen a ReadyForQuery message?
+			if typ == pgwirebase.ServerMsgReady {
+				atomic.StoreInt32(&f.atomics.isServerMsgReadyReceived, 1)
+			}
+
+			// When ForwardMsg gets blocked on Read, we will unblock that by
+			// closing serverConn through f.Close().
+			if _, err := f.serverInterceptor.ForwardMsg(f.clientConn); err != nil {
+				return err
+			}
+
+		case stateTransferInProgress:
+			if err := f.runTransferWithTimeout(
```

Comment:
idea: Instead of implementing the forwarding goroutines as a state machine, initiating the transfer could ask both of the goroutines to terminate. That allows us to implement transfer the same way authenticate is implemented. As a single function and goroutine. Once the transfer is canceled or completed, the goroutine managing the transfer could restart the forwarding goroutines.

IMO the main advantage to implementing it by stopping the forwarding goroutines is we can keep the steady state code simple and we can group all of the transfer logic together.